### PR TITLE
1162 Reorder operations correctly in commit endpoint

### PR DESCRIPTION
**Pull request by Wauplin**
Fix https://github.com/huggingface/huggingface_hub/issues/1162.

**With this PR:**
- Order of the operations in the commit payload is the same as the input of `create_commit(...)`. It was not the case before which causes trouble if someone wants to delete then create a file.
- If 2 operations are modifying the same file, a warning is issued. Only case without warning is "delete a file then add this file" or "delete folder then add file that would have been in this folder".
- `CommitOperationAdd` is validated at initialization (no need for `.validate()` or `._upload_info()` calls anymore)

**Warning:** I made some breaking changes in `validate_preupload_info`, `fetch_upload_modes` and `prepare_commit_payload`. I did not care too much as those 3 methods are defined in a private module and used exclusively for the in `create_commit` . I checked for safety and it's not used anywhere in `diffusers`, `transformers`, `skops` and `datasets`.

**Potential issue:** if someone adds twice the same file, once a LFS content, once a regular content then the final upload mode will be the one from the second operation. Might cause issue with request payload or gitaly. Anyway, this is a weird case so let's not care of the user has an error. A warning message is triggered before upload so the user should notice the problem. (note: this issue was already existing before this PR)

(cc @severo who raised the issue)

### CODE REVIEW

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -1909,8 +1910,8 @@ def create_commit(
             num_threads=num_threads,
         )
         commit_payload = prepare_commit_payload(
-            additions=additions_with_upload_mode,
-            deletions=deletions,
```
Comment (coyotte508) [inline src/huggingface_hub/hf_api.py:1913]: Is the local `deletions` variable still useful?

**Review by coyotte508 — APPROVED** [review]
conceptual ACK

File: src/huggingface_hub/hf_api.py (hunk 1 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:1913]: Yup true we don't really need it anymore. Only remaining case was to count this list but we don't need the `deletions` list to be actually defined to count them.
I made the change, thanks :+1:

File: tests/test_commit_api.py [hunk 2]
```diff
@@ -46,3 +50,49 @@ def test_explicit_folder(self):
     def test_is_folder_wrong_value(self):
         with self.assertRaises(ValueError):
             CommitOperationDelete(path_in_repo="path/to/folder", is_folder="any value")
+
+
+class TestWarnOnOverwritingOperations(unittest.TestCase):
```
Comment (LysandreJik) [inline tests/test_commit_api.py:55]: Clean class!

**Review by LysandreJik — APPROVED** [review]
Very nice, thanks!

**Comment by Wauplin** [comment]
Thanks for the reviews @coyotte508 @LysandreJik ! I'm merging :)

**Pull request merged**